include/ginkgo/core/distributed/neighborhood_communicator.hpp

```diff
@@ -0,0 +1,134 @@
+    explicit neighborhood_communicator(communicator base);
+
+    /**
+     * Create a neighborhood_communicator from an index map.
+     *
+     * The receive neighbors are defined by the remote indices and their
+     * owning ranks of the index map. The send neighbors are deduced
+     * from that through collective communication.
+     *
+     * @tparam LocalIndexType  the local index type of the map
+     * @tparam GlobalIndexType  the global index type of the map
+     * @param base  the base communicator
+     * @param imap  the index map that defines the communication pattern
+     */
+    template <typename LocalIndexType, typename GlobalIndexType>
+    neighborhood_communicator(
+        communicator base,
+        const distributed::index_map<LocalIndexType, GlobalIndexType>& imap);
+
+    /**
+     * Create a neighborhood_communicator by explicitly defining the
+     * neighborhood lists and sizes/offsets.
+     *
+     * @param base  the base communicator
+     * @param sources  the ranks to receive from
+     * @param recv_sizes  the number of elements to recv for each source
+     * @param recv_offsets  the offset for each source
+     * @param destinations  the ranks to send to
+     * @param send_sizes  the number of elements to send for each destination
+     * @param send_offsets  the offset for each destination
+     */
+    neighborhood_communicator(
+        communicator base,
+        const std::vector<distributed::comm_index_type>& sources,
+        const std::vector<comm_index_type>& recv_sizes,
+        const std::vector<comm_index_type>& recv_offsets,
+        const std::vector<distributed::comm_index_type>& destinations,
+        const std::vector<comm_index_type>& send_sizes,
+        const std::vector<comm_index_type>& send_offsets);
+
+    /**
+     * Communicate data from all ranks to all other ranks using the
+     * neighboorhood communication MPI_Ineighbor_alltoallv. See MPI
+     * documentation for more details
+     *
+     * @param exec  The executor, on which the message buffers are located.
+     * @param send_buffer  the buffer to send
+     * @param send_type  the MPI_Datatype for the send buffer
+     * @param recv_buffer  the buffer to gather into
+     * @param recv_type  the MPI_Datatype for the recv buffer
+     *
+     * @return  the request handle for the call
+     */
+    request i_all_to_all_v(std::shared_ptr<const Executor> exec,
+                           const void* send_buffer, MPI_Datatype send_type,
+                           void* recv_buffer,
+                           MPI_Datatype recv_type) const override;
+
+    std::unique_ptr<collective_communicator> create_with_same_type(
+        communicator base,
+        const distributed::index_map_variant& imap) const override;
+    /**
+     * Creates the inverse neighborhood_communicator by switching sources
+     * and destinations.
+     *
+     * @return  collective_communicator with the inverse communication pattern
+     */
```

Comment:
Probably better to just copy doc from the base class and add the details there ?